libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -1245,7 +1245,16 @@ bool ContractCompiler::visit(ForStatement const& _forStatement)
 
 	// for's loop expression if existing
 	if (_forStatement.loopExpression())
+	{
+		Arithmetic previousArithmetic = m_context.arithmetic();
+		if (
+			*_forStatement.annotation().isSimpleCounterLoop &&
+			m_optimiserSettings == OptimiserSettings::standard()
```

Comment:
Hmm... making this dependent on having *all* the optimizations enabled seems suboptimal. And totally non-obvious - at the very least this fact should be documented. 

Not sure how to do it better though since the `enabled` flag does not exist internally. Should we maybe do this if any of the optimizer flags is enabled instead?

Reply:
After looking at how the optimizer settings are handled and configured in the compiler by `CommandLineParser` class, I changed it to check that `m_optimiserSettings != OptimiserSettings::none()`. I think that makes the compiler enable the `simpleCounterLoop` property (when adequate) if there is any optimization requested.

Reply:
About this: can you add a new optimizer setting for that instead?
It should just be default of the default settings for an enabled optimizer (so for the CLI interface ``--optimize`` enables it and there doesn't have to be any finer control for disabling it) and for standard json, it could also be an implied default of ``optimizer { enabled: true }``, but can be more finely controled as ``optimizer { enabled: true, details: { simpleCounterLoopUncheckedIncrement: false } }``. That would make it behave very similarly to e.g. the peephole optimizer and should be quick to do and much cleaner.

This would involve:
- Extending the `OptimiserSettings` with a new field (defaulting to ``false``, but turned on in ``standard()`` and ``full()``).
- Parsing the setting in ``StandardCompiler`` (search for ``"peephole"`` as a reference)
- Exposing the setting in the `else` branch within `CompilerStack::createMetadata`

Reply:
> Exposing the setting in the else branch within CompilerStack::createMetadata (to make sure it can be properly set for source verification)

As an aside here - it is very easy to forget about this big if unaware and a common pitfall. I've seen this in multiple PRs, and sometimes been bitten by it myself, though I think in the end we always caught it. Still, maybe we should think about a mechanism that would make forgetting this more obvious. Either some automatic mechanism to include all settings by default (so that you can see it in tests and exclude if wrong) or some assert that will fail if a setting is not accounted for.

Reply:
A nightly CI run similar to bytecode comparison that compares direct compilation with recompilation on the basis of metadata should already help with that...